libsolidity/experimental/analysis/TypeInference.cpp

```diff
@@ -204,31 +129,34 @@ bool TypeInference::visit(TypeClassDefinition const& _typeClassDefinition)
 		_typeClassDefinition.typeVariable().accept(*this);
 	}
 
-	std::map<std::string, Type> functionTypes;
-
 	solAssert(m_analysis.annotation<TypeClassRegistration>(_typeClassDefinition).typeClass.has_value());
 	TypeClass typeClass = m_analysis.annotation<TypeClassRegistration>(_typeClassDefinition).typeClass.value();
+
 	Type typeVar = m_typeSystem.typeClassInfo(typeClass).typeVariable;
 	auto& typeMembersAnnotation = annotation().members[typeConstructor(&_typeClassDefinition)];
 
+	std::map<std::string, Type> const& functionTypes = m_analysis.annotation<TypeClassMemberRegistration>().typeClassFunctions.at(typeClass);
+
 	for (auto subNode: _typeClassDefinition.subNodes())
 	{
-		subNode->accept(*this);
 		auto const* functionDefinition = dynamic_cast<FunctionDefinition const*>(subNode.get());
 		solAssert(functionDefinition);
-		// TODO: need polymorphicInstance?
-		auto functionType = polymorphicInstance(typeAnnotation(*functionDefinition));
-		if (!functionTypes.emplace(functionDefinition->name(), functionType).second)
-			m_errorReporter.fatalTypeError(3195_error, functionDefinition->location(), "Function in type class declared multiple times.");
+
+		// For type class members the type is assigned by TypeClassMemberRegistration.
+		// Fill in the `type` annotation to avoid `visit(FunctionDefinition)` giving it a fresh one.
+		Type functionType = functionTypes.at(functionDefinition->name());
+		solAssert(!annotation(*functionDefinition).type.has_value());
+		annotation(*functionDefinition).type = functionType;
```

Comment:
This is a bit awkward and I'm starting to doubt if this is really what I was supposed to do. What I did here is to create the function type for every type class member in `TypeClassMemberRegistration` and then copy it into this annotation here during `TypeInference`.

I wonder if it would make more sense to have a separate pass that initializes `type` annotations with fresh variables instead so that other passes before type inference can refer to them.

I also expected to be able to move more of the stuff from `visit(TypeClassDefinition)` but it seems to be mostly tied to unification so most of it had to stay and I still have iterate over `FunctionDefinitions`.

The notes say this:

> Pass 3: visit type class bodies, collecting function signatures (can involve defined types, so depends on 2!)

The thing is, we're not really collecting signatures - we don't know these until after inference - before it we only have two fresh variables, one for the whole tuple of arguments and one for a tuple of returns. And doing this does not require knowing anything about defined types since those are just two opaque type variables before inference. So I could just as well do this already in `TypeClassRegistration`.

I must be missing something because I don't really see why exactly this needs to be a separate pass.

Reply:
How do the signatures of type class functions depend on type inference? (They don't...)

Reply:
Can you clarify what you mean by function signatures here? Technically they don't depend on inference but before inference all you know is that it's some underspecified function type - its inputs and outputs are just fresh variables. I was assuming that signature is the fully inferred type.

Actually, initially I assumed that it's what stated in the declaration in the source file (however incomplete it may be), but this does not seem like collecting that would be of any use.